src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncStateHandler.scala

```diff
@@ -0,0 +1,446 @@
+    }
+  }
+
+  def discardLastBlocks(startBlock: BigInt, blocksToDiscard: Int): Unit = {
+    (startBlock to ((startBlock - blocksToDiscard) max 1) by -1).foreach { n =>
+      blockchain.getBlockHeaderByNumber(n).foreach { headerToRemove =>
+        blockchain.removeBlock(headerToRemove.hash, withState = false)
+      }
+    }
+    appStateStorage.putBestBlockNumber((startBlock - blocksToDiscard - 1) max 0)
+  }
+
+  def requestNodes(from: Peer)(syncState: SyncState): (RequestNodes, SyncState) = {
+    val (nonMptNodesToGet, remainingNonMptNodes) = syncState.pendingNonMptNodes.splitAt(syncConfig.nodesPerRequest)
+    val (mptNodesToGet, remainingMptNodes) = syncState.pendingMptNodes.splitAt(syncConfig.nodesPerRequest - nonMptNodesToGet.size)
+    val nodesToGet = nonMptNodesToGet ++ mptNodesToGet
+
+    (RequestNodes(from, nodesToGet, mptNodesToGet, nonMptNodesToGet), syncState.copy(
+      pendingNonMptNodes = remainingNonMptNodes,
+      pendingMptNodes = remainingMptNodes
+    ))
+  }
+
+  def requestHeaders(from: Peer)(syncState: SyncState): (RequestHeaders, SyncState) = {
+    val numberOfHeaders: BigInt =
+    if (syncConfig.blockHeadersPerRequest < (syncState.safeDownloadTarget - syncState.bestBlockHeaderNumber))
+      syncConfig.blockHeadersPerRequest
+    else
+      syncState.safeDownloadTarget - syncState.bestBlockHeaderNumber
+
+    (RequestHeaders(from, syncState.bestBlockHeaderNumber + 1, numberOfHeaders), syncState)
+  }
+
+  def requestBodies(from: Peer)(syncState: SyncState): (RequestBodies, SyncState) = {
+    val (blockBodiesToGet, remainingBlockBodies) = syncState.blockBodiesQueue.splitAt(syncConfig.blockBodiesPerRequest)
+    (RequestBodies(from, blockBodiesToGet), syncState.copy(blockBodiesQueue = remainingBlockBodies))
+  }
+
+  def requestReceipts(from: Peer)(syncState: SyncState): (RequestReceipts, SyncState) = {
+    val (receiptsToGet, remainingReceipts) = syncState.receiptsQueue.splitAt(syncConfig.receiptsPerRequest)
+    (RequestReceipts(from, receiptsToGet), syncState.copy(receiptsQueue = remainingReceipts))
+  }
+
+  private def handleMptNode(syncState: SyncState, mptNode: MptNode): Seq[HashType] = mptNode match {
+    case n: LeafNode =>
+      import io.iohk.ethereum.network.p2p.messages.PV63.AccountImplicits._
+      //if this fails it means that we have leaf node which is part of MPT that do not stores account
+      //we verify if node is paert of the tree by checking its hash before we call handleMptNode() in line 44
+      val account = Try(n.value.toArray[Byte].toAccount) match {
+        case Success(acc) => Some(acc)
+        case Failure(e) =>
+          log.debug(s"Leaf node without account, error while trying to decode account ${e.getMessage}")
+          None
+      }
+
+      val evm = account.map(_.codeHash)
+      val storage = account.map(_.storageRoot)
+
+      blockchain.saveNode(ByteString(n.hash), n.toBytes, syncState.targetBlock.number, withSnapshotSave = false)
```

Comment:
maybe you could create helper method because this saveNode param configuration is mainly dependent on `n` and `synsState` values